s2core/src/main/scala/com/kakao/s2graph/core/Graph.scala

```diff
@@ -319,11 +319,20 @@ object Graph {
 
       val successCallback = (kvs: util.ArrayList[KeyValue]) => {
         val edgeWithScores = Edge.toEdges(kvs, queryParam, prevScore, isInnerCall = false, parentEdges)
-        QueryResult(q, stepIdx, queryParam, edgeWithScores)
+        val sample = queryParam.sample
+        val sampledEdgesWithScores = if (sample >= 0 ) {
+          // sample only from non-degree edges
+          val (degreeEdges, edges) = edgeWithScores.partition { case (edge, _) => edge.propsWithTs.contains(LabelMeta.degreeSeq) }
+          val sampledEdges = Random.shuffle(edges).take(sample)
```

Comment:
Hi, IMHO would you better to use draw k samples using RNG from populations. It much faster than shuffle-take, the more population goes bigger. 
like this..
'''
# draw k samples from n-elements array
if n < small_value:
    return naive way (does not matter)
else if n > big_enough:
    selected = set()
    while True:
        nth = RNG.next()  % n  # choose RNG library carefully for both performance randomness 
        if nth not in selected:
             selected.add(nth)
        if len(selected) >= k:
             break
    return selected.aslist()
'''

even though it's more complicated then shuffle-take... I think it would be worth consideration if you are worrying about the large size of samples.